### PR TITLE
[REF] base: modernise ir.ui.view

### PR DESCRIPTION
**Pull request by Julien00859**
Aggregation of all the changes related to [Task 2463632](https://www.odoo.com/web#id=2463632&model=project.task&view_type=form&cids=1&menu_id=5195) and its sub-tasks.

### [IMP] base: fields_view_get() benchmark

Because our task is performance-related, we added a way to quickly benchmark the changes. We used py-spy to generate a speedscope (because the new `profile` was not merged yet) out of the benchmark and measure the functions but the new profiler can be used instead. All the commit that are mentioning performance improvements have been realized thanks to this benchmark with the following modules installed: `website_sale,mrp,contacts,crm,timesheet`.

py-spy cli:

    py-spy record -F -r 200 --format speedscope -o speedscope.json -- python3 odoo-bin -d <yourdatabase> --test-tags render_all_views --stop-after-init

To use the new built-in profiler (see #66590) instead:

     diff --git a/odoo/addons/base/tests/test_views.py b/odoo/addons/base/tests/test_views.py
     index b04fdd981bb..05b0c2da617 100644
     --- a/odoo/addons/base/tests/test_views.py
     +++ b/odoo/addons/base/tests/test_views.py
     @@ -3094,6 +3094,7 @@ class TestRenderAllViews(common.TransactionCase):

          @common.users('demo', 'admin')
          def test_render_all_views(self):
     +      with self.profile():
              env = self.env(context={'lang': 'en_US'})
              count = 0
              elapsed = 0

###  [REF] base: introduce view._get_combined_arch()

Also: [IMP] all: read_combined(['arch']['arch'] => get_combined_arch()

The only public API available to retrieve a combined view arch is `read_combined()` which is a `read` function where the `arch` field contains the combined views as a string.

We suggest to provide a way to get the combined architecture outside of a read. If possible, there should be a way to retrieve the combined arch either as string either as a low level etree.Element object. There are codes that re-parse the combined view in a etree Element and that would benefit to get an etree Element right away.

Two new methods were added:

* `get_combined_arch()` takes no argument and returns the combined arch as string
* `_get_combined_arch()` takes no argument and returns the combined arch as an etree Element.

Because the `read_combined` function was basically only used to get the combined arch (`read_combined(['arch'])['arch']` almost everywhere), we adapted all modules to use either of the two new method instead and deprecated `read_combined`.

### [REF] base: view hierarchy retrieval and combination

The current view combination mechanism involves multiple steps: (i) find all the primary views that are ancestors to the current view, for each primary view (starting with the oldest one) (ii) combine the primary view (iii) find all its extensions, (iv) order them in combination order, (v) combine them.

_Example_

In the hierarchy below, the views (1, 4) marked with a * are primary views, the view (6) marked with a ! is the view we want to combine.

              1*
             / \
            2   3
           / \
          4*  5
         /   / \
        6!  7   8

The primary views are ordered as follow (first to combine first): 1, 4.
The views are ordered as follow (first to combine first): 1, 2, 5, 7, 8, 3, 4, 6.

The current implementation uses a recursive approach: for every primary view, it queries the database to retrieve its descendants and merge them together. We argue that the recursion is sub-efficient because it queries the database more. We also argue that it is difficult to work with the current implementation and suggest to refactor it.

The view methods `_get_combined_arch()`, `get_inheriting_views_arch()`, `apply_view_inheritance()`, `_apply_view_inheritance()` and were responsible to combine the views together:

* `_get_combined_arch()`: implemented the above-mentioned recursion, it was responsible to find the nearest primary view ancestor, delegate the combination of the primary view's parent view thanks to a recursive _get_combined_arch() call.
* `get_inheriting_views_arch()`: big fat sql query to find all the extensions of the current primary view
* `apply_view_inheritance()`: build the above-mentioned hierarchy (only with 1 primary view)
* `_apply_view_inheritance()`: traverse in-order depth-first the hierarchy to combine the views

We changed the structure as follow:

* `_get_combined_arch()`: find all the views that are parent to the current view
* `_get_inheriting_views()`: find all the descendants for all the views
* `_get_combined_arch()` (again): build the above-mentioned hierarchy (with all primary views)
* `_combine()`: traverse the hierarchy in a clever order to combine the views

### Other commits

They are primary optimizations, please refer to the commit message of each commit.

### CODE REVIEW

**Review by Xavier-Do — CHANGES_REQUESTED** [review]

File: addons/website/models/ir_ui_view.py
Comment (rco-odoo) [inline addons/website/models/ir_ui_view.py:314]: This is not correct: the code expects a recordset, but the super-call returns a dict instead.
I think it would be easier to stick to the former convention, i.e., return views as a recordset.
This makes it easier for overridings.
Building the `hierarchy` tree in `_get_combined_arch()` instead is not an issue, IMHO.

**Comment by tde-banana-odoo** [comment]
Target 14 par défaut, on se fait souvent avoir :D

**Comment by Julien00859** [comment]
Nan, j'avais un `rebase -i <viewhash>` dans mon historique shell, j'ai rebase sur master du coup full commits se sont injectés entre ce viewhash et la nouvelle master. J'ai rebase -i et voyant ma connerie j'ai voulu cancel mais git rebase a préréfé me mettre en commiter de tous les commits intermédiaire :+1:

File: odoo/addons/base/tests/test_views.py
Comment (Julien00859) [inline odoo/addons/base/tests/test_views.py:3093]: The `--test-file` argv ignores the `-standard` tag. How could we make `--test-file` skips this test or is there a `--test-tags` equivalent that we could use to run all tests classes in a module ?

**Comment by Julien00859** [comment]
Hello @odoo/rd-security :wave: I changed the `safe_eval` function so it accepts a `frozendict` type as globals dict so I could memoize the evaluation

**Comment by odony** [comment]
> Hello @odoo/rd-security wave I changed the `safe_eval` function so it accepts a `frozendict` type as globals dict so I could memoize the evaluation

Could your PR description give a good high-level description of what the PR is supposed to achieve, in terms of goals and implementation choices? You're modifying a critical part of the framework, and many people are going to revisit this PR in the future. The task/subtasks seem to mostly contain draft ideas / bare titles.

Maybe everything is in the commit messages and comments and you just need to build a summary out of it. That would be helpful for reviewers. :pray:

File: odoo/addons/base/models/ir_ui_view.py
Comment (Julien00859) [inline odoo/addons/base/models/ir_ui_view.py:849]: stack_level=2

File: odoo/addons/base/tests/test_views.py
Comment (xmo-odoo) [inline odoo/addons/base/tests/test_views.py:3111]: That's not a test, why is that a test? Also using timeit just to get aggregate times seems to be rather missing the point.

File: odoo/addons/base/models/ir_ui_view.py
Comment (xmo-odoo) [inline odoo/addons/base/models/ir_ui_view.py:2023]: Just do that unconditionally?
```suggestion
self.fields_get = Model.with_context(lang=None).fields_get()
```
?

And the commit message is overly verbose and providing informations we probably don't care about, I'd only keep the last two paragraphs:

> `NameManager`'s validation passes should not be language-dependent. Fetching translated model & view information thus incurs additional costs for no resaon.

> Ensure translations are never enabled for `NameManager`'s work, reducing its cost from ~13% to ~8.5% of view rendering.

Also would need some information as to e.g. which view & setup was involved here.

**Review by xmo-odoo — COMMENTED** [review]
There needs to be a pass through the commit messages, there are issues of orthography and grammar e.g.

> flamegraphes

or 

> recursive calls to not render very well

and the repetitive "Part of an effort to refactor ir.ui.view." seems unnecessary and redundant, both as part of this here PR and once merged.

But some (many?) messages are also muddled / unclear e.g. 

>  [REF] base: introduce view._get_combined_arch()

> Part of an effort to refactor ir.ui.view.
>
> The `view.read_combined` public method only returns the combined architecture as string. It parses the XML source files in a `etree.Element` merge them and return a stringyfied version of the arch.
>
> Sometimes, the caller needs to perform some operation on the XML nodes so it re-parses the output of the function. The multiple serialisations and de-serialisations are causing performance issues.
>
> We introduce two new functions: `view.get_combined_arch()` and `view._get_combined_arch()`. The two combine all the views the current one is depending on. The first (public) one returns the combined arch as string. The second (private) one returns the combined arch as a `etree.Element`.

What `read_combined` does doesn't matter that much here, we can probably infer it. The main point of interest is the purpose of `_get_combined_arch`, and aside from unspecified callers the message shoudl mention *`read_combined` itself does a parsing & serialization dance internally*. So I'd go with something along the lines of:

> Extract architecture combination out of `read_combined` to work on etree.Element exclusively: since `read_combined` returns a serialized `arch`, its own recursive calls as well as callers needing to post-process the architecture have to re-parse the `arch` then serialize it back.
>
> This is wasteful and can be avoided by providing the arch as an `Element`, which is what `_get_combined_arch` does.

The message should also explain what the purpose of `get_combined_arch` is, is it just convenience for tests and the one weird case of auth_totp_portal? And why should it be available over RPC? That seems completely unnecessary.

File: odoo/addons/base/models/ir_ui_view.py
Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py:2023]: `with_context` creates a new environment, it is cheap to create a new env for not free. Because we instantiate many NameManagers, we gain a few CPU cycles by only changing the context when it is needed. @Xavier-Do suggested we remove the lang from the context one for all earlier but I couldn't figure how to do so.

I'll change the commit message to include the list of modules by DB had during the benchmark.

File: odoo/addons/base/tests/test_views.py
Comment (Julien00859) [reply odoo/addons/base/tests/test_views.py:3110]: The other solution is to create a sub-command, IMO a test-case `-standard` is fine. Before I added timeit, we had a bare `for _ in range(5)` loop to render each view multiple times and generate better speedscopes. Thanks to timeit, I can also get the precise time spent only in the function and log it. Timeit is nice, both for speedscope (left-heavy) and for the logs.

Comment (xmo-odoo) [reply odoo/addons/base/tests/test_views.py:3110]: > Thanks to timeit, I can also get the precise time spent only in the function and log it. Timeit is nice, both for speedscope (left-heavy) and for the logs.

You can trivially get the same using precise timers.

The point of timeit is to more easily aggregate calls in order to hide variance and reach a "relevant" total runtime (0.2s using the CLI / `autorange`, pyperf is a bit more complicated but wants a runtime of at least 0.1s) but that's not at all how it's being used here.

File: odoo/addons/base/models/ir_ui_view.py
Comment (xmo-odoo) [reply odoo/addons/base/models/ir_ui_view.py:2023]: > `with_context` creates a new environment, it is cheap to create a new env for not free. Because we instantiate many NameManagers, we gain a few CPU cycles by only changing the context when it is needed.

If you're doing that kind of useless micro-optimisations, then you should also skip the context update when `lang` is already `None`, which you're not.

File: odoo/addons/base/tests/test_views.py
Comment (Julien00859) [reply odoo/addons/base/tests/test_views.py]: You're right, let's just use `time.perf_counter()`

File: odoo/addons/base/models/ir_ui_view.py
Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py]: I rerun the benchmark, `if Model.env.context.get('lang')` is a micro-optimisation compared with `with_context(lang=False)`, went for the simpler solution to always remove the lang.

File: odoo/addons/base/models/ir_ui_view.py
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_view.py:826]: Is `encode()` actually necessary?

Comment (rco-odoo) [reply odoo/addons/base/models/ir_ui_view.py:826]: If it is, `'utf-8'` isn't, as it is the default value for `encode()`.

Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py:826]: IIRC the word "string" in "fromstring" refers to py2's string aka bytes in py3.

**Comment by Julien00859** [comment]
@odoo/rd-security https://github.com/odoo/odoo/pull/71419/commits/a1fdae8d90db66b83ce215914651b21eef432ccc J'ai fait mon possible pour mieux documenter, est-ce que comme ça c'est mieux ?

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [reply odoo/addons/base/models/ir_ui_view.py:826]: > IIRC the word "string" in "fromstring" refers to py2's string aka bytes in py3.

The [lxml](https://lxml.de/api/lxml.etree-module.html#fromstring) doc is not fancy, but you could just try it, and you'll see it accepts both text and bytes, so no need for extra encoding.

```py
Python 3.8.5 (default, May 27 2021, 13:30:53) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from lxml import etree
>>> etree.fromstring("<a></a>")
<Element a at 0x7f1c894b3ec0>
```

**Comment by Julien00859** [comment]
| Commit | Title | Admin (s) | Demo (s) | Runtime (s) | Query count (k) |
| ------ | ----- | --------- | -------- | ----------- | --------------- |
| f6caa4d171e | Baseline: benchmark | 2.805 | 2.206 | 30.54 | 31.576 |
| c1c32255461 | get_combined_arch: base | 2.801 | 2.185 | 30.33 | 31.486 |
| 13f8796cdff | get_combined_arch: all | 2.698 | 2.09 | 29.17 | 28.601 |
| 17a8e70a47d | Hierarchy | 2.569 | 1.994 | 28.07 | 26.076 |
| 4b6ad4f8747 | Some Fixes | 2.554 | 1.964 | 27.58 | 26.076 |
| 496bdd78ee2 | Memoization | 2.548 | 1.96 | 27.8 | 26.076 |
| cc37560ad5a | Flatten | 2.571 | 1.979 | 28.01 | 26.076 |
| caae654b1a5 | NameManager | 2.407 | 1.844 | 24.8 | 23.3 |

ir_ui_view_refactor_bench

**Comment by Julien00859** [comment]
@rco-odoo you had concern about the memo, I logged in a file every string that we `safe_eval` in `safedb` and `literal_eval` in `evaldb`. Here are the aggregated results: https://bin.readthedocs.fr/offitc.sh, this confirm the `CacheInfo` of yersterday, ~95% hit rate for `safe_eval`, ~30% hit rate for `literal_eval`.

**Comment by xmo-odoo** [comment]
Questions wrt the cache commit:

* could it make sense to rejigger the arguments to `safe_eval` in order to pass in a constant if there is no context?
* I'd expect most attributes (not attrs) to be static values (e.g. `True`, `1`, ...), maybe special-casing them to bypass evalution entirely would be even better?
* I'd expect `attrs` to be extremely repetitive *within* a view but not so much *between* views (the low hit rate seems to support this assumption), maybe the cache could be a simple `dict` used on a per-view basis rather than a cross-view cache? I'd also assume `literal_eval` to be much faster (than `safe_eval`) so the overhead of the cache would have a higher impact

**Comment by Julien00859** [comment]
>     * I'd expect most attributes (not attrs) to be static values (e.g. `True`, `1`, ...), maybe special-casing them to bypass evalution entirely would be even better?

You're right:

	~/Odoo $ wc -l safedb
	3349 safedb
	~/Odoo $ grep -E -e "^\('(0|1|True|False)'" safedb | wc -l
	3311

Special-casing them would be better, I'm doing it in a few minutes...

>     * I'd expect `attrs` to be extremely repetitive _within_ a view but not so much _between_ views (the low hit rate seems to support this assumption), maybe the cache could be a simple `dict` used on a per-view basis rather than a cross-view cache? I'd also assume `literal_eval` to be much faster (than `safe_eval`) so the overhead of the cache would have a higher impact

The overhead of the cache indeed standout in my benchmarks but the memoization was still beneficial, using a cache per model would be beneficial.

	~/Odoo $ sort evaldb | uniq -c | sort -n | cut -d '(' -f 1 | trim | uniq -c
	    171 1
	    795 2
	     13 3
	    179 4
	      1 5
	     36 6
	     21 8
	      3 9
	      4 10
	     10 12
	      3 14
	      3 16
	      1 28
	      1 90

* "795 2" means that 795 different expressions appeared twice
* "1 90" means that 1 expression appeared 90 times, the expression is `"{'invisible': [('active', '=', True)]}"`

On a total of 3300 expressions, only 3129 benefit from the cache and 1539 benefit multiple times.

The two files if you want to run your own analysis: [evaldb.txt](https://github.com/odoo/odoo/files/6638345/evaldb.txt), [safedb.txt](https://github.com/odoo/odoo/files/6638346/safedb.txt).

I suggest we move the Memoization to a new pull-request so it does not block this one too long. What do you think ?

**Comment by xmo-odoo** [comment]
> I suggest we move the Memoization to a new pull-request so it does not block this one too long. What do you think ?

I don't mind either. We could also merge with just the special-casing of the attribute values and explore changes to the literal bits later, as long as they don't make things *slower* they're a gain regardless.

**Comment by Julien00859** [comment]
See b872cd86c381

**Comment by Julien00859** [comment]
@robodoo rebase-ff

File: odoo/addons/base/models/ir_ui_view.py
Comment (rco-odoo) [inline odoo/addons/base/models/ir_ui_view.py:637]: Maybe we should also fill in the cache of `mode`, since that field will be used in `_combine()`.

**Comment by rco-odoo** [comment]
@robodoo delegate+

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py:603]: without braces this is incorrect if `where_clause` has more than one condition

```suggestion
                WHERE id IN %s AND ({where_clause})
```

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py:610]: same thing
```suggestion
                      AND ({sub_where_clause})
```

Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py:610]: There are parenthesis already, see expression.py L614

```where_clause: '((FALSE OR "ir_ui_view"."website_id" IS NULL) AND ("ir_ui_view"."active" = %s))'```

File: odoo/addons/base/models/ir_ui_view.py
Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py:603]: Same thing

Comment (rco-odoo) [reply odoo/addons/base/models/ir_ui_view.py:603]: Actually, `expression` wraps every boolean operator with a lot of parentheses
https://github.com/odoo/odoo/blob/14.0/odoo/osv/expression.py#L609-L618

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py:596]: Ouch :see_no_evil:  Can't you use `where_clause` unmodified in both places inside the full query? The `WITH RECURSIVE` and `UNION` are different scopes so there should be no conflict.

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [reply odoo/addons/base/models/ir_ui_view.py:610]: > There are parenthesis already, see expression.py L614

The braces are an implementation detail that is not guaranteed at all by `get_sql()`, so we *should* include extra braces. It's trivial to obtain a `where_clause` that isn't surrounded by braces via `get_sql()`, whereas there is no problem with having an extra pair of braces.

I understand that it's not your code, but it should have been caught in earlier reviews.

That's safer, defensive programming, and we want the framework to be solid and defensive.

Comment (Julien00859) [reply odoo/addons/base/models/ir_ui_view.py]: done

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py:79]: `vs`? `value_str` perhaps?

File: odoo/addons/base/models/ir_ui_view.py
Comment (odony) [inline odoo/addons/base/models/ir_ui_view.py:596]: I believe we need to keep this check even if `_get_inheriting_views()` is private, as it is reachable from other public methods and the raw SQL is not otherwise protected.

e.g. I can now call `get_combined_arch()` in RPC without being an admin, and I get the view arch

**Comment by Julien00859** [comment]
@robodoo r+

**Pull request closed without merging**
